pkg/template/processor_test.go

```diff
@@ -0,0 +1,604 @@
+			"kind": "RoleBinding",
+			"metadata": map[string]interface{}{
+				"name": "rb1",
+			},
+		},
+	}
+	rb2 := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"kind": "RoleBinding",
+			"metadata": map[string]interface{}{
+				"name": "rb2",
+			},
+		},
+	}
+
+	t.Run("filter namespaces", func(t *testing.T) {
+
+		t.Run("return one", func(t *testing.T) {
+			// given
+			objs := []runtime.RawExtension{
+				{
+					Object: ns1,
+				},
+				{
+					Object: rb1,
+				},
+			}
+			// when
+			result := template.Filter(objs, template.RetainNamespaces)
+			// then
+			require.Len(t, result, 1)
+			assert.Equal(t, ns1, result[0].Object)
+
+		})
+		t.Run("return multiple", func(t *testing.T) {
+			// given
+			objs := []runtime.RawExtension{
+				{
+					Object: ns1,
+				},
+				{
+					Object: rb1,
+				},
+				{
+					Object: ns2,
+				},
+				{
+					Object: rb2,
+				},
+			}
+			// when
+			result := template.Filter(objs, template.RetainNamespaces)
+			// then
+			require.Len(t, result, 2)
+			assert.Equal(t, ns1, result[0].Object)
+			assert.Equal(t, ns2, result[1].Object)
+		})
+		t.Run("return none", func(t *testing.T) {
+			// given
+			objs := []runtime.RawExtension{
+				{
+					Object: rb1,
+				},
+				{
+					Object: rb2,
+				},
+			}
+			// when
+			result := template.Filter(objs, template.RetainNamespaces)
+			// then
+			require.Empty(t, result)
+		})
+	})
+
+	t.Run("filter other resources", func(t *testing.T) {
+
+		t.Run("return one", func(t *testing.T) {
+			// given
+			objs := []runtime.RawExtension{
+				{
+					Object: ns1,
+				},
+				{
+					Object: rb1,
+				},
+			}
+			// when
+			result := template.Filter(objs, template.RetainAllButNamespaces)
+			// then
+			require.Len(t, result, 1)
+			assert.Equal(t, rb1, result[0].Object)
+
+		})
+		t.Run("return multiple", func(t *testing.T) {
+			// given
+			objs := []runtime.RawExtension{
+				{
+					Object: ns1,
+				},
+				{
+					Object: rb1,
+				},
+				{
+					Object: ns2,
+				},
+				{
+					Object: rb2,
+				},
+			}
+			// when
+			result := template.Filter(objs, template.RetainAllButNamespaces)
+			// then
+			require.Len(t, result, 2)
+			assert.Equal(t, rb1, result[0].Object)
+			assert.Equal(t, rb2, result[1].Object)
+		})
+
+		t.Run("return none", func(t *testing.T) {
+			// given
+			objs := []runtime.RawExtension{
+				{
+					Object: ns1,
+				},
+				{
+					Object: ns2,
+				},
+			}
+			// when
+			result := template.Filter(objs, template.RetainAllButNamespaces)
+			// then
+			require.Empty(t, result)
+		})
+	})
+}
+
+func TestProcessAndApply(t *testing.T) {
+
+	t.Run("should create namespace alone", func(t *testing.T) {
+
+		// given
+		s := addToScheme(t)
+		namespace, commit, user := templateVars()
+		values := paramsKeyValues(namespace, commit, user)
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s)
+
+		// when
+		objs, err := p.Process(templateContent(namespaceObj), values)
+		require.NoError(t, err)
```

Comment:
it's a detail, but these two lines should belong to "given" part - the test verifies only the "apply" functionality... 
but as I said - it's a detail ;-)

Reply:
fair point, addressed in 718d67c